[1730] add mcb rollover command

misaka opened the pull request:
### Context

Need to copy the things.

### Changes proposed in this pull request

Copy the things.

### Guidance to review

Currently based off of 1564-attach-recruitment-cycle-to-provider. Once that branch is merged to master, this should be rebased onto master.

### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally

fofr commented on app/models/course.rb:
```diff
@@ -312,6 +312,36 @@ def to_s
+            vac_status: SiteStatus.default_vac_status_given(
+              study_mode: new_course.study_mode
+            ),
+            status: :new_status
```
We need to edit `applications_accepted_from` too.

misaka replied:
Do we know what it gets set to?

benilovj commented on lib/mcb/commands/providers/rollover.rb:
```diff
@@ -0,0 +1,18 @@
+summary "Create a copy of provider's courses for the next recruitment cycle"
+
```
It'd be helpful to hint at the optional provider codes by defining the usage string. I had to read the code to work it out.

timabell commented:
![Selection_687](https://user-images.githubusercontent.com/19378/60984919-69ca4000-a334-11e9-9e31-6df55ba9db39.png)

I see in the 2020 provider:

* year_code stuck at 2019
* no `organisation_provider` records
* no `provider_enrichment` records

defong commented:
> Don't forget to update accrediting provider references

Be careful, it can get recursive, 
accrediting provider can be a provider as well.

Ie, 

a `Hayes Uni` is a `Provider` as it has a `Course`
a `SCITT` is a  `Provider` as it has a `Course` and it's `Course` is accredited by `Hayes Uni`

timabell commented:
![Selection_688](https://user-images.githubusercontent.com/19378/60986497-7dc37100-a337-11e9-96c7-edd79fc63054.png)

Things I notice on a course

* start date still 2019
* accrediting provider still linking back to 2019 copy of provider
* no `course_site` link
* no `course_subject` link

timabell commented:
Before / after numbers
![Selection_689](https://user-images.githubusercontent.com/19378/60986855-499c8000-a338-11e9-8e8a-84449a5ae3c1.png)

timabell commented:
`provider_ucas_preference` has `provider_id` but is missing the FK, and hasn't been duplicated

defong commented:
> Before / after numbers
Does that makes sense?

![image](https://user-images.githubusercontent.com/470137/60987162-efe88580-a338-11e9-8584-94464d55defd.png)

timabell commented:
@defong I think it's fine for audit to record the additions. The zero is because it's wiped during sanitizing.

defong commented:
Do we need to worry about `PdgeCourse`? @benilovj it has `provider_code` and `course_code` as the dataset...

defong commented:
hmmm i wonder how the audit thing works... maybe a simple sanity check... audit.count = `course.count + provider.count + site.count` etc

defong commented on spec/lib/mcb/commands/providers/rollover_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      course_code: course.course_code
+    )
+    expect(new_course).not_to be_nil
+    expect(new_course).not_to eq course
```
maybe `course.attributes` equals `new_course.attributes` minus `id` etc

defong commented on spec/models/site_spec.rb:
```diff
@@ -79,4 +79,47 @@
+      it 'does not make a copy of the site' do
+        # Something strange is going on with sites ... setting the code as we
+        # do for next_site doesn't seem to work so we have to assign it here.
+        next_site.update code: site.code
```
maybe along here... https://github.com/DFE-Digital/manage-courses-backend/blob/63ab44c1ee1f5b3567edbf9de0e1fee9728fab2c/app/models/site.rb#L29

misaka replied:
Ya, I saw that but I couldn't quite fit it in with the behaviour I was seeing, which is why I just dropped it. I'm tempted to leave it for now.

defong commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add mcb command to roll-over courses and providers](https://trello.com/c/qmkpvWoH/1730-add-mcb-command-to-roll-over-courses-and-providers)

misaka commented:
> > Don't forget to update accrediting provider references
> 
> Be careful, it can get recursive,
> accrediting provider can be a provider as well.

Ya, this is tricky. Because we don't have the singular provider record which spans recruitment cycles, perhaps `provider_id` should be `provider_code`. Otherwise we're in dragon territory ... do we duplicate the accrediting provider when we encounter it on the course?

misaka commented:
> Do we need to worry about `PdgeCourse`? @benilovj it has `provider_code` and `course_code` as the dataset...

https://github.com/DFE-Digital/manage-courses-backend/blob/63ab44c1ee1f5b3567edbf9de0e1fee9728fab2c/app/models/pgde_course.rb#L10-L13

Looks like we don't have to worry about this one.

benilovj commented:
> > > Don't forget to update accrediting provider references
> > 
> > 
> > Be careful, it can get recursive,
> > accrediting provider can be a provider as well.
> 
> Ya, this is tricky. Because we don't have the singular provider record which spans recruitment cycles, perhaps `provider_id` should be `provider_code`. Otherwise we're in dragon territory ... do we duplicate the accrediting provider when we encounter it on the course?

Can’t the mcb command just do a second pass of the courses in the new cycle and update their accredited bodies?

misaka commented:
> > > > Don't forget to update accrediting provider references
> > > 
> > > 
> > > Be careful, it can get recursive,
> > > accrediting provider can be a provider as well.
> > 
> > 
> > Ya, this is tricky. Because we don't have the singular provider record which spans recruitment cycles, perhaps `provider_id` should be `provider_code`. Otherwise we're in dragon territory ... do we duplicate the accrediting provider when we encounter it on the course?
> 
> Can’t the mcb command just do a second pass of the courses in the new cycle and update their accredited bodies?

Sure. As a hack. I don't like it. :) It opens up the possibility of linking to an accredited provider in another recruitment cycle.

defong commented:
> > Do we need to worry about `PdgeCourse`? @benilovj it has `provider_code` and `course_code` as the dataset...
> 
> https://github.com/DFE-Digital/manage-courses-backend/blob/63ab44c1ee1f5b3567edbf9de0e1fee9728fab2c/app/models/pgde_course.rb#L10-L13
> 
> Looks like we don't have to worry about this one.

Going going gone?
https://github.com/DFE-Digital/manage-courses-backend/pull/596

misaka commented:
So just to address the gaps presented here:

```
| public.course_site | 31412 | 45108 |
| public.site        | 8256  | 13071 |
```

These are caused by invalid sites that don't have anything for their `address3` field. We've dropped the validation to solve (#605)

```
| public.course_subject | 39183 | 39183 |
```

Addressed in commit cae140d2

```
| public.organisation_provider    | 1201  | 1201  |
| public.provider_ucas_preference | 1701  | 1701  |
```

Addressed in commit 5e6250f

```
| public.provider_enrichment | 2037  | 2037  |
| public.organisation_user   | 19275 | 19275 |
```

These don't need to be copied.

misaka commented:
Here's the latest counts:

```
|                            | current cycle | next cycle |
| providers                  |          1899 |       1899 |
| providers.organisations    |          1201 |       1201 |
| providers.ucas_preferences |          1701 |       1701 |
| courses                    |         17705 |      17705 |
| courses.subjects           |         39183 |      39183 |
| sites                      |          8257 |       7819 |
```

So `sites` is better, but there's still a gap. I wonder if it's another validation, will have to look.

These were calculated with commands:

```ruby
RecruitmentCycle.has_many :organisations, through: :providers
RecruitmentCycle.has_many :ucas_preferences, through: :providers
RecruitmentCycle.has_many :subjects, through: :courses

current = RecruitmentCycle.first
next_recycle = RecruitmentCycle.last

current.providers.count
next_recycle.providers.count

current.courses.count
next_recycle.courses.count

current.sites.count
next_recycle.sites.count

current.organisations.count
next_recycle.organisations.count

current.ucas_preferences.count
next_recycle.ucas_preferences.count

current.subjects.count
next_recycle.subjects.count
```

misaka commented:
It looks like there are a further 441 sites that are still invalid, due either `address1`, `postcode` and `location_name` being blank.

defong commented:
in order of likelihood, `location_name`, `address1`, `postcode` aka `Main site`

timabell commented:
updated numbers

prod > rolled-over
![Selection_693](https://user-images.githubusercontent.com/19378/61209657-41ef2980-a6f2-11e9-951b-c7d8bc3dadab.png)

misaka commented:
```
| public.course_site | 55116 | 31412 |
```

This discrepancy is because we don't copy over suspended or discontinued sites. This has been confirmed with @fofr.

misaka commented:
@timabell Could you re-review pls?

defong commented on app/models/course.rb:
```diff
@@ -324,6 +324,27 @@ def to_s
+
+      new_provider.courses << new_course
+
+      if (last_enrichment = enrichments.latest_first.first)
```
just copy the lastest, which can be either `published` or `draft`

defong commented on app/models/course_enrichment.rb:
```diff
@@ -110,6 +110,15 @@ def unpublish(initial_draft: true)
+  def copy_to_course(new_course)
+    new_enrichment = self.dup
+    if new_enrichment.published?
+      new_enrichment.last_published_timestamp_utc = nil
```
you want to nil it for draft as well

a draft with last published is a subsequent draft

this will most likely lead to frontend displaying it wrong

misaka replied:
Done.

defong commented on app/models/site.rb:
```diff
@@ -53,6 +53,30 @@ def to_s
+    unless new_site
+      new_site = self.dup
+      new_site.provider_id = new_provider.id
+      new_site.save(validate: false)
```
is this bit still needed, after #605

misaka replied:
Ya, it's still needed. `postcode` and `address1` are still blank on some records. 😿

defong commented on spec/models/course_spec.rb:
```diff
@@ -769,4 +769,134 @@
     end
   end
+
+  describe '#copy_to_provider' do
```
would be nice to have tests that covers

`new_course` 
`content_status` should be either `draft` or `empty` covers the enrichments
`ucas_status` should be `new` or `not running` (if original is also not running) covers the site statuses
`open_for_applications` should be nil

my thinking is along the lines that a rollovered course will have the state as listed above as those exact states will be presented to user on frontend

misaka commented:
@timabell Ttttiiiiimmmmm!!

timabell commented:
all ur builds are failing

fofr commented:
A summary of a conversation with @misaka. When running the command we're seeing some courses showing as being "Published with unpublished changes" in the next cycle:

![Screen Shot 2019-07-18 at 10 28 47](https://user-images.githubusercontent.com/319055/61446412-ff725a80-a946-11e9-868e-3cf24d6b0ec4.png)

At this point nothing in the next cycle will have been published, so this is a bug. It's because we're copying the published enrichment as well as the most recent draft – the status logic sees that it's been published before and shows the wrong status.

We won't ever use these published enrichments – we will not be using them to automatically publish when we do the cycle flip in October. The only way content can get onto Find is if the provider publishes their course manually.

This means we can fix this bug by only copying the draft enrichment.

defong commented:
> This means we can fix this bug by only copying the draft enrichment.
I think we covered it yesterday with...
https://github.com/DFE-Digital/manage-courses-backend/pull/589/files/79c5fc80f7d6629eadbf0f03b4edcbeea85b01ca#discussion_r304319810
